Remove repeated section and some edits (#918)
timescale/docs: timescaledb/how-to-guides/compression/backfill-historical-data.md

# Backfill historical data on compressed chunks
You can backfill historical data by temporarily decompressing any compressed
chunks. Backfilling means inserting data with older timestamps, beyond the
recent past. Depending on your compression policy, the chunks corresponding to
these timestamps might already be compressed.

For help with backfilling on compressed chunks, you can use the backfilling
function in the [TimescaleDB extras][timescaledb-extras] repository. You can
also manually backfill without the function.

<highlight type="note">
Backfilling is useful for inserting batches of data, rather than individual
rows. For individual rows, use a normal `INSERT`. TimescaleDB [handles the
compression behind the scenes](https://www.timescale.com/blog/timescaledb-2-3-improving-columnar-compression-for-time-series-on-postgresql/).
</highlight>

## Backfill with the backfilling function

> You can backfill data by using the `decompress_backfill` function. This

Reviewer: ```suggestion
You can backfill data by using the `decompress_backfill` stored procedure. This
```

function:

>     * Halts the compression policy
>     * Identifies the chunks where the backfilled data belongs
>     * Decompresses the chunks
>     * Inserts data from the temporary table into the hypertable
>     * Re-enables the compression policy

Author: ```suggestion
1.    Halts the compression policy
2.   Identifies the chunks where the backfilled data belongs
3.   Decompresses the chunks
4.   Inserts data from the temporary table into the hypertable
5.   Re-enables the compression policy
```

> Before you begin, add the [`decompress_backfill` function and its supporting
> functions][timescaledb-extras-backfill] to your database.

Reviewer: ```suggestion
Before you begin, add the [`decompress_backfill` stored procedure and its supporting
functions][timescaledb-extras-backfill] to your database.
```

> Before you begin, add the [`decompress_backfill` stored procedure and its supporting
> functions][timescaledb-extras-backfill] to your database.

Author: ```suggestion
Before you begin, add the [`decompress_backfill` stored procedure and its supporting
functions][timescaledb-extras-backfill] to your database.
You can copy and paste the procedure and function declarations from `backfill.sql` and
run them in your database.
```

compression behind the scenes](https://www.timescale.com/blog/timescaledb-2-3-improving-columnar-compression-for-time-series-on-postgresql/).
</highlight>

## Backfill with the backfilling function
You can backfill data by using the `decompress_backfill` function. Before you
begin, add the [`decompress_backfill` function and its supporting
functions][timescaledb-extras-backfill] to your database.

<procedure>

## Backfilling with the backfilling function

<highlight type="note">
Backfill data by first inserting it into a temporary table. Temporary tables are
short-lived. They are automatically dropped when the database session ends. No
clean-up is required. If you backfill regularly, you can choose to use a normal
table instead. This allows multiple writers to insert data into the table at the
same time. When backfilling is complete, clean up by truncating or dropping your
table.
</highlight>

1.  At the `psql` prompt, create a table with the same schema as the hypertable
    you're backfilling into. In this example, the hypertable is named `cpu`.
    ```sql
    CREATE TEMPORARY TABLE cpu_temp AS SELECT * FROM cpu WITH NO DATA;
    ```

> 

Reviewer 2: ```suggestion
```

1.  Insert your data into the backfill table.

> 

Reviewer 2: ```suggestion
```

1.  Call the supplied backfill function. This function:
      * Halts the compression policy
      * Identifies the chunks where the backfilled data belongs
      * Decompresses the chunks
      * Inserts data from the temporary table into the hypertable

>       * Re-enables the compression policy

Reviewer 2: This reads as concept info, can we move it out of the procedure?

    ```sql
    CALL decompress_backfill(staging_table=>'cpu_temp', destination_hypertable=>'cpu');
    ```

</procedure>

## Backfill manually

> If you prefer not to use the `decompress_backfill` function, you can perform its

Reviewer 2: ```suggestion
If you prefer not to use the `decompress_backfill` function, you can perform the
```

> steps manually. In this procedure, you turn off your compression policy and

Reviewer 2: ```suggestion
steps manually. In this procedure, you turn off the compression policy and
```

> identify your compressed chunks before inserting your backfilled data.

Reviewer 2: ```suggestion
identify your compressed chunks before inserting the backfilled data.
```

<procedure>

### Backfilling manually
1.  At the `psql` prompt, find the `job_id` of the compression policy:
    ```sql
    SELECT s.job_id
    FROM timescaledb_information.jobs j
      INNER JOIN timescaledb_information.job_stats s ON j.job_id = s.job_id

>       WHERE j.proc_name = 'policy_compression' AND s.hypertable_name = <TARGET_TABLE>;

Reviewer: ```suggestion
      WHERE j.proc_name = 'policy_compression' AND s.hypertable_name = <TARGET_TABLE>;
```
I'd really like to get some input here from the larger team. I'm tired enough tonight that I honestly cannot think of what I'd want, but this is one of the few places I recall seeing the tag-like placeholders for variable or parameter names, and I just don't think it works.

In the old docs we appear to have mixed things, which isn't better, but in context we tend to use sample table/parameter values. I understand it's not "that much" clearer, but the correct way to substitute parameters in any language is always tricky IMO.

Maybe a cursory  look at what some other similar docs/SQL do?

Author: Thanks for the input, Ryan. I did a cursory review of what other industry documentation uses:

|Docs|Formal-ish syntax definitions|Sample values|Tag-like placeholders|Notes|
|-|-|-|-|-|
|PostgreSQL|✅|✅||PostgreSQL is not the best example to learn from, because their docs don't have how-to guides that I've found. They're mainly API references, which start with a formal-ish syntax definition and end with examples. This is a good pattern for our references but not as useful in context for a task or concept.|
|InfluxDB||✅|✅|Uses a tag-like placeholder immediately followed by an example|
|Stripe|✅||✅|Leans towards sample values in their how-to guides. Placeholders are rare.|
|Clickhouse||✅|||
|Snowflake||✅|✅|Leans towards tag-like placeholders|
|MongoDB||✅||Uses example variables but heavily defines their examples, like literally "Imagine a pizza store receives two orders..."|

In our recent discussion on this topic, we decided on placeholder parameters as the primary method with examples when needed, but maybe we should flip that, given sample variables seem to be more popular? 

I do think some examples need to be better defined, for example if we were to replace `<JOB_ID>` below with the actual job id, it wouldn't be clear at a quick glance what that string of numbers is supposed to be, so it would require more explanatory text

Reviewer 2: I probably am leaning towards the Influx method, but it's going to require more words generally, and a concerted effort to go through and update everything.

Reviewer 3: I like the Snowflake approach only because we may not have to do a lot of rewrites.

    ```
1.  Pause compression to prevent the policy from trying to compress chunks that
    you're currently working on:
    ``` sql
    SELECT alter_job(<JOB_ID>, scheduled => false);
    ```
1.  Decompress the chunks that you want to modify:
    ``` sql
    SELECT decompress_chunk('_timescaledb_internal._hyper_2_2_chunk');
    ```

>     Repeat for each chunk. Alternatively, you can decompress a set of chunks

Reviewer 2: Can we move the alternative option here to a new optional step instead?

    based on a time range using [`show_chunks`][show_chunks]:
    ``` sql
    SELECT decompress_chunk(i) 
      FROM show_chunks(
        '<TARGET_TABLE>', 
        newer_than => '<NEWER_THAN_DATE>', 
        older_than => '<OLDER_THAN_DATE>'
      ) i;
    ```
1.  When you have decompressed all the chunks you want to modify, perform the
    `INSERT` or `UPDATE` commands to backfill the data.
1.  Restart the compression policy job. The next time the job runs, it
    recompresses any chunks that were decompressed.
    ``` sql
    SELECT alter_job(<JOB_ID>, scheduled => true);
    ```
    To recompress chunks immediately, use the `run_job` command:
    ``` sql
    CALL run_job(<JOB_ID>);
    ```

</procedure>

## Future Work [](future-work)

> 

Reviewer 2: ```suggestion
```

</highlight>

## Backfill with the backfilling function
You can backfill data by using the `decompress_backfill` function. This
function:
    * Halts the compression policy
    * Identifies the chunks where the backfilled data belongs
    * Decompresses the chunks
    * Inserts data from the temporary table into the hypertable

>     * Re-enables the compression policy

Reviewer 2: Should these be ordered?

Author: Probably!

## Backfill with the backfilling function
You can backfill data by using the `decompress_backfill` function. This
function:

> 1.   Halts the compression policy
> 2.   Identifies the chunks where the backfilled data belongs
> 3.   Decompresses the chunks
> 4.   Inserts data from the temporary table into the hypertable
> 5.   Re-enables the compression policy

Reviewer 2: ```suggestion
1.   Halts the compression policy
1.   Identifies the chunks where the backfilled data belongs
1.   Decompresses the chunks
1.   Inserts data from the temporary table into the hypertable
1.   Re-enables the compression policy
```

</procedure>

## Backfill manually
If you prefer not to use the `decompress_backfill` function, you can perform the

> steps manually. In this procedure, you turn off the compression policy and

Reviewer 2: ```suggestion
steps manually. In this procedure, turn off the compression policy and
```
